cli/listener/src/main.rs

```diff
@@ -0,0 +1,256 @@
+    output_mode: Option<OutputMode>,
+
+    /// Unique name for identifying
+    #[clap(long, default_value = "solido")]
+    pool: String,
+
+    /// Poll frequency in seconds, defaults to 5 minutes.
+    #[clap(long, default_value = "300")]
+    poll_frequency_seconds: u32,
+
+    /// Location of the SQLite DB file.
+    #[clap(long, default_value = "listener.db")]
+    db_path: String,
+}
+
+struct State {
+    pub solido: Lido,
+}
+
+impl State {
+    pub fn new(
+        config: &mut SnapshotConfig,
+        solido_program_id: &Pubkey,
+        solido_address: &Pubkey,
+    ) -> Result<Self, SnapshotError> {
+        let solido = config.client.get_solido(solido_address)?;
+        Ok(State { solido })
+    }
+}
+
+#[derive(Debug)]
+pub struct ExchangeRate {
+    /// Id of the data point.
+    id: i32,
+    /// Unix timestamp when the data point was logged.
+    timestamp: chrono::DateTime<chrono::Utc>,
+    /// Slot when the data point was logged.
+    slot: Slot,
+    /// Epoch when the data point was logged.
+    epoch: Epoch,
+    /// Pool identifier, e.g. for Solido would be "solido".
+    pool: String,
+    /// Token_a price.
+    token_a: u64,
+    /// Token_b price.
+    token_b: u64,
+}
+
+pub fn create_db(conn: &Connection) -> rusqlite::Result<()> {
+    // The timestamp is stored in ISO-8601 format.
+    conn.execute(
+        "CREATE TABLE IF NOT EXISTS exchange_rate (
+                id          INTEGER PRIMARY KEY,
+                timestamp   TEXT,
+                slot        INTEGER NOT NULL,
+                epoch       INTEGER NOT NULL,
+                pool        TEXT NOT NULL,
+                token_a     INTEGER NOT NULL,
+                token_b     INTEGER NOT NULL
+            )",
+        [],
+    )?;
+    Ok(())
+}
+
+pub struct IntervalPrices {
+    t0: chrono::DateTime<chrono::Utc>,
+    t1: chrono::DateTime<chrono::Utc>,
+    epoch0: Epoch,
+    epoch1: Epoch,
+    price0_lamports: Rational,
+    price1_lamports: Rational,
+}
+
+impl IntervalPrices {
+    pub fn duration_wall_time(&self) -> chrono::Duration {
+        self.t1 - self.t0
+    }
+    pub fn duration_epochs(&self) -> u64 {
+        self.epoch1 - self.epoch0
+    }
+    pub fn growth_factor(&self) -> std::result::Result<Rational, ArithmeticError> {
+        self.price1_lamports / self.price0_lamports
+    }
+    pub fn annual_growth_factor(&self) -> f64 {
+        let year = chrono::Duration::days(365);
+        self.growth_factor()
+            .expect("Overflow happened when calculating growth factor.")
+            .to_f64()
+            .powf(year.num_seconds() as f64 / self.duration_wall_time().num_seconds() as f64)
+    }
+    pub fn annual_percentage_rate(&self) -> f64 {
+        self.annual_growth_factor().mul_add(100.0, -100.0)
+    }
+}
+
+pub fn get_apy_for_period(
+    conn: &Connection,
+    opts: &Opts,
+    from_time: chrono::DateTime<chrono::Utc>,
+    to_time: chrono::DateTime<chrono::Utc>,
+) -> rusqlite::Result<Option<f64>> {
```

Comment:
We should return the `IntervalPrices` here, and delay converting to a float as far as possible. A float itself requires an out-of-band description of what time period the APY is measured over. But the `IntervalPrice` type is self-descriptive: if you receive a value of type `IntervalPrice`, and that is all you have, you know what it means. If you receive an `f64`, even if you know it’s an APY, you don’t know over what period.